### ISSUE TITLE
xla_spawn and run_language_modeling slow on TPUs

### ISSUE THREAD

**Issue by ksjae** [huggingface/transformers#7962]
## Environment info

- `transformers` version: 3.4.0
- Platform: Linux-4.9.0-13-amd64-x86_64-with-debian-9.13
- Python version: 3.6.10
- PyTorch version (GPU?): 1.8.0a0+e5ed037 (False)
- Tensorflow version (GPU?): not installed (NA)
- Using GPU in script?: No (but using TPU)
- Using distributed or parallel set-up in script?: using xla_spawn.py

### Who can help
@LysandreJik @sgugger 
or the writer of examples/language-modeling/run_language_modeling.py or a TPU master

## Information

Model I am using (Bert, XLNet ...): GPT2

The problem arises when using:
* [ ] the official example scripts: (give details below)
* [x] my own modified scripts: examples/language-modeling/run_language_modeling.py but with HF datasets

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [x] my own task or dataset: Text generation

## To reproduce

1. Modify examples/language-modeling/run_language_modeling.py to below
```
import logging
import math
import os
import glob
import datasets
from dataclasses import dataclass, field
from typing import Optional

from datasets import list_datasets, load_dataset

from transformers import (
    CONFIG_MAPPING,
    MODEL_WITH_LM_HEAD_MAPPING,
    AutoConfig,
    AutoModelWithLMHead,
    AutoTokenizer,
    DataCollatorForLanguageModeling,
    DataCollatorForPermutationLanguageModeling,
    HfArgumentParser,
    LineByLineTextDataset,
    PreTrainedTokenizer,
    TextDataset,
    Trainer,
    TrainingArguments,
    set_seed,
)


logger = logging.getLogger(__name__)


MODEL_CONFIG_CLASSES = list(MODEL_WITH_LM_HEAD_MAPPING.keys())
MODEL_TYPES = tuple(conf.model_type for conf in MODEL_CONFIG_CLASSES)


@dataclass
class ModelArguments:
    """
    Arguments pertaining to which model/config/tokenizer we are going to fine-tune, or train from scratch.
    """

    model_name_or_path: Optional[str] = field(
        default=None,
        metadata={
            "help": "The model checkpoint for weights initialization. Leave None if you want to train a model from scratch."
        },
    )
    model_type: Optional[str] = field(
        default=None,
        metadata={"help": "If training from scratch, pass a model type from the list: " + ", ".join(MODEL_TYPES)},
    )
    config_name: Optional[str] = field(
        default=None, metadata={"help": "Pretrained config name or path if not the same as model_name"}
    )
    tokenizer_name: Optional[str] = field(
        default=None, metadata={"help": "Pretrained tokenizer name or path if not the same as model_name"}
    )
    cache_dir: Optional[str] = field(
        default=None, metadata={"help": "Where do you want to store the pretrained models downloaded from s3"}
    )


@dataclass
class DataTrainingArguments:
    """
    Arguments pertaining to what data we are going to input our model for training and eval.
    """

    train_data_file: Optional[str] = field(
        default=None, metadata={"help": "The input training data file (a text file)."}
    )
    eval_data_file: Optional[str] = field(
        default=None,
        metadata={"help": "An optional input evaluation data file to evaluate the perplexity on (a text file)."},
    )
    line_by_line: bool = field(
        default=False,
        metadata={"help": "Whether distinct lines of text in the dataset are to be handled as distinct sequences."},
    )

    mlm: bool = field(
        default=False, metadata={"help": "Train with masked-language modeling loss instead of language modeling."}
    )
    mlm_probability: float = field(
        default=0.15, metadata={"help": "Ratio of tokens to mask for masked language modeling loss"}
    )
    plm_probability: float = field(
        default=1 / 6,
        metadata={
            "help": "Ratio of length of a span of masked tokens to surrounding context length for permutation language modeling."
        },
    )
    max_span_length: int = field(
        default=5, metadata={"help": "Maximum length of a span of masked tokens for permutation language modeling."}
    )

    block_size: int = field(
        default=-1,
        metadata={
            "help": "Optional input sequence length after tokenization."
            "The training dataset will be truncated in block of this size for training."
            "Default to the model max input length for single sentence inputs (take into account special tokens)."
        },
    )
    overwrite_cache: bool = field(
        default=False, metadata={"help": "Overwrite the cached training and evaluation sets"}
    )
    arrow: bool = field(
        default=True,
        metadata={
            "help": "Use Arrow-based HF NLP for optimization."
        },
    )


def get_dataset(
    args: DataTrainingArguments,
    tokenizer: PreTrainedTokenizer,
    evaluate: bool = False,
    cache_dir: Optional[str] = "./cache",
):
    tokenizer.pad_token = "<|endoftext|>"
    tokenizer._pad_token = "<|endoftext|>"
    #tokenizer.pad_token_id = 50256
    file_path = args.eval_data_file if evaluate else args.train_data_file
    if True:
        dataset = datasets.load_from_disk(file_path)
        dataset.set_format(type='torch', columns=['input_ids'])
        return dataset
    
    if False:
        dataset = load_dataset("text", data_files=[file_path], split='train')
        dataset = dataset.map(lambda ex: tokenizer(ex["text"], add_special_tokens=True,
                                                truncation=True, max_length=args.block_size), batched=True)
        dataset.set_format(type='torch', columns=['input_ids'])
        dataset.save_to_disk(file_path+'.arrow')
        return dataset
    
    if args.line_by_line:
        return LineByLineTextDataset(tokenizer=tokenizer, file_path=file_path, block_size=args.block_size)
    else:
        return TextDataset(
            tokenizer=tokenizer,
            file_path=file_path,
            block_size=args.block_size,
            overwrite_cache=args.overwrite_cache,
            cache_dir=cache_dir,
        )
    """
    dataset = load_dataset("text", data_files=file_path, split="train")
    dataset = dataset.map(lambda ex: tokenizer(ex["text"], add_special_tokens=True,
                                            truncation=True, max_length=args.block_size), batched=True)
    dataset.set_format(type='torch', columns=['input_ids'])
    return dataset
    """

def main():
    # See all possible arguments in src/transformers/training_args.py
    # or by passing the --help flag to this script.
    # We now keep distinct sets of args, for a cleaner separation of concerns.

    parser = HfArgumentParser((ModelArguments, DataTrainingArguments, TrainingArguments))
    model_args, data_args, training_args = parser.parse_args_into_dataclasses()

    if data_args.eval_data_file is None and training_args.do_eval:
        raise ValueError(
            "Cannot do evaluation without an evaluation data file. Either supply a file to --eval_data_file "
            "or remove the --do_eval argument."
        )

    if (
        os.path.exists(training_args.output_dir)
        and os.listdir(training_args.output_dir)
        and training_args.do_train
        and not training_args.overwrite_output_dir
    ):
        raise ValueError(
            f"Output directory ({training_args.output_dir}) already exists and is not empty. Use --overwrite_output_dir to overcome."
        )

    # Setup logging
    logging.basicConfig(
        format="%(asctime)s - %(levelname)s - %(name)s -   %(message)s",
        datefmt="%m/%d/%Y %H:%M:%S",
        level=logging.INFO if training_args.local_rank in [-1, 0] else logging.WARN,
    )
    logger.warning(
        "Process rank: %s, device: %s, n_gpu: %s, distributed training: %s, 16-bits training: %s",
        training_args.local_rank,
        training_args.device,
        training_args.n_gpu,
        bool(training_args.local_rank != -1),
        training_args.fp16,
    )
    logger.info("Training/evaluation parameters %s", training_args)

    # Set seed
    set_seed(training_args.seed)

    # Load pretrained model and tokenizer
    #
    # Distributed training:
    # The .from_pretrained methods guarantee that only one local process can concurrently
    # download model & vocab.

    if model_args.config_name:
        config = AutoConfig.from_pretrained(model_args.config_name, cache_dir=model_args.cache_dir)
    elif model_args.model_name_or_path:
        config = AutoConfig.from_pretrained(model_args.model_name_or_path, cache_dir=model_args.cache_dir)
    else:
        config = CONFIG_MAPPING[model_args.model_type]()
        logger.warning("You are instantiating a new config instance from scratch.")

    if model_args.tokenizer_name:
        tokenizer = AutoTokenizer.from_pretrained(model_args.tokenizer_name, cache_dir=model_args.cache_dir)
    elif model_args.model_name_or_path:
        tokenizer = AutoTokenizer.from_pretrained(model_args.model_name_or_path, cache_dir=model_args.cache_dir)
    else:
        raise ValueError(
            "You are instantiating a new tokenizer from scratch. This is not supported, but you can do it from another script, save it,"
            "and load it from here, using --tokenizer_name"
        )

    tokenizer.pad_token = "<|endoftext|>"
    tokenizer._pad_token = "<|endoftext|>"

    if model_args.model_name_or_path:
        model = AutoModelWithLMHead.from_pretrained(
            model_args.model_name_or_path,
            from_tf=bool(".ckpt" in model_args.model_name_or_path),
            config=config,
            cache_dir=model_args.cache_dir,
        )
    else:
        logger.info("Training new model from scratch")
        model = AutoModelWithLMHead.from_config(config)

    model.resize_token_embeddings(len(tokenizer))

    if config.model_type in ["bert", "roberta", "distilbert", "camembert"] and not data_args.mlm:
        raise ValueError(
            "BERT and RoBERTa-like models do not have LM heads but masked LM heads. They must be run using the"
            "--mlm flag (masked language modeling)."
        )

    if data_args.block_size <= 0:
        data_args.block_size = tokenizer.max_len
        # Our input block size will be the max possible for the model
    else:
        data_args.block_size = min(data_args.block_size, tokenizer.max_len)

    # Get datasets

    train_dataset = (
        get_dataset(data_args, tokenizer=tokenizer, cache_dir=model_args.cache_dir) if training_args.do_train else None
    )
    eval_dataset = (
        get_dataset(data_args, tokenizer=tokenizer, evaluate=True, cache_dir=model_args.cache_dir)
        if training_args.do_eval
        else None
    )
    if config.model_type == "xlnet":
        data_collator = DataCollatorForPermutationLanguageModeling(
            tokenizer=tokenizer,
            plm_probability=data_args.plm_probability,
            max_span_length=data_args.max_span_length,
        )
    else:
        data_collator = DataCollatorForLanguageModeling(
            tokenizer=tokenizer, mlm=data_args.mlm, mlm_probability=data_args.mlm_probability
        )

    # Initialize our Trainer
    trainer = Trainer(
        model=model,
        args=training_args,
        data_collator=data_collator,
        train_dataset=train_dataset,
        eval_dataset=eval_dataset,
        prediction_loss_only=True,
    )

    # Training
    if training_args.do_train:
        model_path = (
            model_args.model_name_or_path
            if model_args.model_name_or_path is not None and os.path.isdir(model_args.model_name_or_path)
            else None
        )
        trainer.train(model_path=model_path)
        trainer.save_model()
        # For convenience, we also re-save the tokenizer to the same directory,
        # so that you can share your model easily on huggingface.co/models =)
        if trainer.is_world_master():
            tokenizer.save_pretrained(training_args.output_dir)

    # Evaluation
    results = {}
    if training_args.do_eval:
        logger.info("*** Evaluate ***")

        eval_output = trainer.evaluate()

        perplexity = math.exp(eval_output["eval_loss"])
        result = {"perplexity": perplexity}

        output_eval_file = os.path.join(training_args.output_dir, "eval_results_lm.txt")
        if trainer.is_world_master():
            with open(output_eval_file, "w") as writer:
                logger.info("***** Eval results *****")
                for key in sorted(result.keys()):
                    logger.info("  %s = %s", key, str(result[key]))
                    writer.write("%s = %s\n" % (key, str(result[key])))

        results.update(result)

    return results


def _mp_fn(index):
    # For xla_spawn (TPUs)
    main()


if __name__ == "__main__":
    main()
```

2. set torch-xla-nightly Conda & set env
3. run script (replace dataset, since I cannot upload 48 GB worth of arrow files)
```
XLA_USE_BF16=1 python3 examples/xla_spawn.py --num_cores 8 examples/language-modeling/train.py --output_dir=kogpt1 --model_type=gpt2 --do_train --train_data_file=/home/ksjcom0705_gmail_com/NEWS_ARROW --overwrite_output_dir --per_device_train_batch_size=6 --save_steps 10000 --num_train_epochs=1 --block_size 2048 --eval_steps 10000 --logging_steps=10000 --tokenizer_name /home/ksjcom0705_gmail_com/kotok tpu_num_cores=8
```

The progress bar will show something like 250s/it, while on GPUs(2 V100s) it's about 1.2 it/s

## Expected behavior

At least give speed similar to GPUs, since my [home-brew code](https://github.com/ksjae/KoGPT2-train) works with similar speed.

Also, MXU utilization is stuck at near-zero.

<img width="2177" alt="image" src="https://user-images.githubusercontent.com/17930170/96816530-663ec780-1458-11eb-9595-be91de708d03.png">

**Comment by LysandreJik** [comment]
The only difference I'm seeing between our script and yours is the data loading, here using `datasets`. Are we aware of TPU slowdown when using `datasets` @lhoestq, @thomwolf?

**Comment by lhoestq** [comment]
I haven't tested `datasets` with TPU yet. I know @sgugger tried once, did you notice slowdowns ?

**Comment by sgugger** [comment]
Not for the new run_glue script introduced in #7917

**Comment by ksjae** [comment]
Is there any metrics/debug information I can provide?

**Comment by phtephanx** [comment]
I have similar issues:
* mxu utilization mostly 0%
* around 150 s/it

setup:
* n1-highmem-16
* TPU v2-8

I tried it both with a map-style dataset using `datasets`' wiki dump and an iterable-style dataset with the trainer adapted. Same result. The slowdown is not on behalf of the data loading but of the forward-pass, loss computation and backpropagation on the tpu.

**Comment by sgugger** [comment]
Are you sure all your batches of inputs have the exact same shape? I tested thoroughly the script on TPUs with the datasets library and:
- when inputs are not all of the same size, the training is excruciatingly slow (which is expected because XLA recompiles the code at each training step in this case)
- when inputs are all of the same size, it runs smoothly and fast.

This is independent of using the datasets library or not, and this is expected behavior on TPUs, as XLA does not handle well dynamic shapes.

**Comment by ksjae** [comment]
How can I check whether inputs are the same size? Or how can I pad the inputs so it has fixed size?

**Comment by sgugger** [comment]
Your dataset is hidden inside the `load_dataset` function, so I can't advise you on how to add padding. There are examples of this in the new `run_mlm.py` script.
As for checking your inputs are all of the same size, it's just a pass through your dataset:
```
shapes = []
for x in dataset:
    shapes.append(x["input_ids"].shape)
print(set(shapes))
```

**Comment by phtephanx** [comment]
Keeping shapes constant fixed the speed issue for me. It went down from 150s/it to few seconds per batch.

Further readings:
* https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#known-performance-caveats
* https://github.com/pytorch/xla/issues/2383
* https://github.com/pytorch/xla/issues/2368

**Comment by ksjae** [comment]
Fixed it, closing.
